lib/gc/Transforms/GPU/GpuToGpuOcl.cpp

```diff
@@ -0,0 +1,560 @@
+        size *= vectorType.getNumElements() / 8;
+      } else {
+        llvm::errs() << "Unsupported type: " << type << "\n";
+        return false;
+      }
+      argSize.emplace_back(helper.idxConstant(rewriter, loc, size));
+    }
+
+    auto array = [&](SmallVector<Value> &values) {
+      auto size = helper.idxConstant(rewriter, loc, values.size());
+      auto arrayPtr = rewriter.create<LLVM::AllocaOp>(loc, helper.ptrType,
+                                                      helper.idxType, size);
+      for (size_t i = 0, n = values.size(); i < n; i++) {
+        auto elementPtr = rewriter.create<LLVM::GEPOp>(
+            loc, helper.ptrType, helper.idxType, arrayPtr,
+            helper.idxConstant(rewriter, loc, i));
+        auto value = values[i];
+        if (auto cast = value.getDefiningOp<UnrealizedConversionCastOp>()) {
+          assert(getConstantIntValue(cast.getOperand(0)));
+          value = helper.idxConstant(
+              rewriter, loc, getConstantIntValue(cast.getOperand(0)).value());
+        }
+        rewriter.create<LLVM::StoreOp>(loc, value, elementPtr);
+      }
+      return arrayPtr.getResult();
+    };
+
+    auto ctx = function.getArgument(0);
+    auto argNum =
+        helper.idxConstant(rewriter, loc, adaptor.getKernelOperands().size());
+    auto createKernelCall = funcCall(
+        rewriter, GPU_OCL_KERNEL_CREATE, helper.ptrType,
+        {helper.ptrType, helper.idxType, helper.ptrType, helper.ptrType,
+         helper.ptrType, helper.ptrType, helper.idxType, helper.ptrType},
+        loc,
+        {ctx, spirvSize, spirv, name, array(gridSize), array(blockSize), argNum,
+         array(argSize)});
+    auto result = createKernelCall.getResult();
+
+    // Save the kernel pointer to the global var using CAS
+    auto null = rewriter.create<LLVM::ZeroOp>(loc, helper.ptrType);
+    auto ptrPtr = rewriter.create<LLVM::AddressOfOp>(loc, ptr);
+    auto casResult = rewriter.create<LLVM::AtomicCmpXchgOp>(
+        loc, ptrPtr, null, result, LLVM::AtomicOrdering::acq_rel,
+        LLVM::AtomicOrdering::monotonic);
+    auto casFlag = rewriter.create<LLVM::ExtractValueOp>(
+        loc, rewriter.getI1Type(), casResult, 1);
+
+    auto body = &function.getBody();
+    auto thenBlock = rewriter.createBlock(body);
+    auto elseBlock = rewriter.createBlock(body);
+    rewriter.setInsertionPointToEnd(&body->front());
+    rewriter.create<LLVM::CondBrOp>(loc, casFlag, thenBlock, elseBlock);
+
+    // Then block
+    rewriter.setInsertionPointToStart(thenBlock);
+    rewriter.create<LLVM::ReturnOp>(loc, result);
+
+    // Else block
+    // The kernel has already been created by another thread, destroying this
+    // one.
+    rewriter.setInsertionPointToStart(elseBlock);
+    helper.destroyKernels(rewriter, loc, result);
+    result = rewriter.create<LLVM::ExtractValueOp>(loc, helper.ptrType,
+                                                   casResult, 0);
+    rewriter.create<LLVM::ReturnOp>(loc, result);
```

Comment:
Not sure I follow, why do we destroy the kernel here?

Reply:
If the current value of the global field is not null, then some other thread has already created and assigned the the value. In this case, we destroy the created kernel and return that one, created by another thread.